File: test-npm-pkg/ts/tests/test-multi-agent.ts
import { pathToFileURL } from 'url';
import { client, runTestSection, logResult, organizationId } from "../utils/utils.js";
import { randomUUID } from "crypto";

/**
 * TEST MULTI-AGENT & SUB-AGENTS (SCENARIO-BASED)
 * Focuses on Sub-agent chat, history, and prompt suggestions.
 */
async function testMultiAgent() {
  console.log("\n" + "=".repeat(70));
  console.log("🚀 STARTING: MULTI-AGENT & SUB-AGENTS TEST");
  console.log("=".repeat(70));

  const ts = Date.now();
  const state = {
    assistantId: null as string | null,
    chatId: null as string | null,
    sessionId: randomUUID(),
    orgId: organizationId as string || "",
  };

  try {
    // --- Section 1: Preparation ---
    await runTestSection("Preparation", async () => {
      const assistants = await client.chatAi.listAssistants();
      if (assistants.length > 0) {
        state.assistantId = assistants[0].id;
      } else {
        const asst = await client.chatAi.createAssistant({
            name: `SDK Sub-Agent Test ${ts}`,
            workflow_name: `sub_agent_test_${ts}`
        });
        state.assistantId = asst.id;
      }
      logResult("Using Assistant", state.assistantId);
    }).catch(e => console.error(`\n❌ Section [Preparation] Failed: ${e.message}`));

    // --- Section 2: Prompt Suggestions ---
    await runTestSection("Prompt Suggestions", async () => {
      const suggestions = await client.aiAgent.getAgentPromptSuggestion(state.assistantId!);
      logResult("Prompt Suggestions", suggestions);
    }).catch(e => console.error(`\n❌ Section [Prompt Suggestions] Failed: ${e.message}`));

    // --- Section 3: Chat v1 Lifecycle ---
    await runTestSection("Chat v1 Lifecycle", async () => {
      const chats = await client.aiAgent.listChats({ organization_id: state.orgId, limit: 1 });
      if (chats.length > 0) {
        const chatId = chats[0].id;
        const chat = await client.aiAgent.getChat(chatId, true);
        logResult("Chat v1 Retrieved", chat.id);
      } else {
        console.log("   ⚠️ No Chat v1 found to test retrieval.");
      }
    }).catch(e => console.error(`\n❌ Section [Chat v1 Lifecycle] Failed: ${e.message}`));

    // --- Section 4: Sub-Agent Chat ---
    await runTestSection("Sub-Agent Chat & History", async () => {
      // 1. Create a client chat first to get a real chatId
      const user = await client.aiAgent.getChatClientUser({ id: `sdk-test-user-${ts}` });
      const clientChat = await client.aiAgent.createClientChat({
        id: randomUUID(),
        assistantId: state.assistantId!,
        organizationId: state.orgId,
        userId: user.id,
        selectedVisibilityType: "public",
        message: {
          id: randomUUID(),
          role: "user",
          content: "Hello Sub-agent",
          createdAt: new Date().toISOString(),
          parts: [{ type: "text", text: "Hello Sub-agent" }]
        }
      });
      state.chatId = clientChat._id || clientChat.id;
      logResult("Chat Created for Sub-agent", state.chatId);

      // 2. Stream Sub-agent Chat
      try {
        const response = await client.aiAgent.streamSubAgentChat({
          assistant_id: state.assistantId!,
          organization_id: state.orgId,
          session_id: state.sessionId,
          chat_id: state.chatId!,
          messages: [{ role: "user", content: "Tell me about yourself" }]
        });
        if (response.status === 200) {
            logResult("Sub-agent Chat Stream", "Started (200 OK)");
        } else {
            console.warn(`   ⚠️ Sub-agent chat stream returned status ${response.status}`);
        }
      } catch (e: any) {
        console.warn(`   ⚠️ Sub-agent chat stream failed: ${e.message}`);
      }

      // 3. Get Sub-agent History
      try {
        const history = await client.aiAgent.getSubAgentHistory({
          session_id: state.sessionId,
          chat_id: state.chatId!
        });
        logResult("Sub-agent History", history);
      } catch (e: any) {
        console.warn(`   ⚠️ Sub-agent history failed: ${e.message}`);
      }
    }).catch(e => console.error(`\n❌ Section [Sub-Agent Chat] Failed: ${e.message}`));

    // --- Section 5: Parquet ---
    await runTestSection("Parquet Data Operations", async () => {
        // Use filename without extension as the server likely appends it
        const parquetFileName = `test-${ts}`; 
        try {
            const res = await client.aiAgent.generateParquet({
                data: [{ name: "Test", value: 123 }],
                fileName: parquetFileName
            });
            logResult("Generated Parquet", res.success || true);

            const files = await client.aiAgent.listParquetFiles();
            logResult("Parquet Files", files.length);

            // Deleting using the same name
            await client.aiAgent.deleteParquetFile(parquetFileName);
            logResult("Deleted Parquet", true);
        } catch (e: any) {
            console.warn(`   ⚠️ Parquet operations failed: ${e.message}`);
        }
    }).catch(e => console.error(`\n❌ Section [Parquet Operations] Failed: ${e.message}`));

    console.log("\n" + "=".repeat(70));
    console.log("✅ MULTI-AGENT TEST COMPLETED SUCCESSFULLY");
    console.log("=".repeat(70));

  } catch (err: any) {
    console.error("\n❌ FATAL TEST ERROR:", err.message);
    throw err;
  } finally {
    console.log("\n[Cleanup] Cleaning up resources...");
    try {
        if (state.chatId) await client.aiAgent.deleteClientChat(state.chatId);
        console.log("   Cleanup finished.");
    } catch (e: any) {
        console.warn("   Cleanup failed:", e.message);
    }
  }
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  testMultiAgent().catch(console.error);
}

export { testMultiAgent };

File: test-npm-pip-pkg/py/tests/test_multi_agent.py
import time
import uuid
from utils.utils import client, run_test_section, log_result, organization_id

def test_multi_agent():
    print("\n🚀 STARTING: MULTI-AGENT & SUB-AGENTS TEST (PY)")

    state = {"assistant_id": None, "chat_id": None, "session_id": str(uuid.uuid4())}

    # 1. Preparation
    def preparation():
        assistants = client.chat_ai.list_assistants()
        if assistants:
            state["assistant_id"] = assistants[0]["id"]
        else:
            ts = int(time.time())
            asst = client.chat_ai.create_assistant({
                "name": f"SDK Sub-Agent Test {ts}",
                "workflow_name": f"sub_agent_test_{ts}"
            })
            state["assistant_id"] = asst["id"]
        log_result("Using Assistant", state["assistant_id"])
    run_test_section("Preparation", preparation)

    # 2. Sub-Agent Chat
    def sub_agent_chat():
        # Create a client chat first
        user = client.ai_agent.get_chat_client_user({"id": f"sdk-py-user-{int(time.time())}"})
        client_chat = client.ai_agent.create_client_chat({
            "id": str(uuid.uuid4()),
            "assistantId": state["assistant_id"],
            "organizationId": organization_id,
            "userId": user["id"],
            "message": {
                "id": str(uuid.uuid4()),
                "role": "user",
                "parts": [{"type": "text", "text": "Hello Sub-agent"}]
            }
        })
        state["chat_id"] = client_chat.get("_id") or client_chat.get("id")
        log_result("Chat Created for Sub-agent", state["chat_id"])

        if state["chat_id"]:
            try:
                res = client.ai_agent.stream_sub_agent_chat({
                    "assistant_id": state["assistant_id"],
                    "organization_id": organization_id,
                    "session_id": state["session_id"],
                    "chat_id": state["chat_id"],
                    "messages": [{"role": "user", "content": "Tell me about yourself"}]
                })
                if res.status_code == 200:
                    log_result("Sub-agent Chat Stream", "Started (200 OK)")
            except Exception as e:
                print(f"   ⚠️ Sub-agent chat stream failed: {str(e)}")

            try:
                history = client.ai_agent.get_sub_agent_history(
                    session_id=state["session_id"],
                    chat_id=state["chat_id"]
                )
                log_result("Sub-agent History", len(history.get("messages", [])))
            except Exception as e:
                print(f"   ⚠️ Sub-agent history failed: {str(e)}")
            
            # Cleanup
            client.ai_agent.delete_client_chat(state["chat_id"])
    run_test_section("Sub-Agent Chat", sub_agent_chat)

    print("\n✅ Multi-Agent Test Completed.")

if __name__ == "__main__":
    test_multi_agent()
